### PR TITLE
rerank 3.5 notebook

### PR DESCRIPTION
**Pull request by jpekmez**
This pull request adds a new notebook to the repository, which demonstrates how to deploy the Cohere-rerank-english model package from the AWS Marketplace using Amazon SageMaker. The notebook includes instructions for subscribing to the model package, creating an endpoint, performing real-time inference, and cleaning up resources.

- The notebook begins with an introduction to Cohere's Large Language Models (LLMs) and their use cases, such as generation, summarization, classification, content moderation, information extraction, semantic search, and contextual entity extraction.
- It then provides pre-requisites for running the notebook, including the need to open it from an Amazon SageMaker Notebook Instance or Amazon SageMaker Studio, and ensuring the IAM role has the required permissions.
- The notebook guides users through subscribing to the model package on the AWS Marketplace, specifying the model package ARN, and creating an endpoint using Boto3.
- It includes code for creating an input payload, performing real-time inference using the `rerank` method, and visualizing the output.
- Finally, the notebook provides instructions for cleaning up resources, including deleting the model and unsubscribing from the listing.

### CODE REVIEW

File: notebooks/sagemaker/Deploy rerank v3.5 model.ipynb [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    "    {\"Title\":\"Customer Support is Busy\",\"Content\":\"Good morning, I have been trying to reach your customer support team for the past week but I keep getting a busy signal. Can you please help me?\"},\n",
+    "    {\"Title\":\"Received Wrong Item\",\"Content\":\"Hi, I have a question about my recent order. I received the wrong item and I need to return it.\"},\n",
+    "    {\"Title\":\"Customer Service is Unavailable\",\"Content\":\"Hello, I have been trying to reach your customer support team for the past hour but I keep getting a busy signal. Can you please help me?\"},\n",
+    "    {\"Title\":\"Return Policy for Defective Product\",\"Content\":\"Hi, I have a question about the return policy for this product. I purchased it a few weeks ago and it is defective.\"},\n",
+    "    {\"Title\":\"Wrong Item Received\",\"Content\":\"Good morning, I have a question about my recent order. I received the wrong item and I need to return it.\"},\n",
+    "    {\"Title\":\"Return Defective Product\",\"Content\":\"Hello, I have a question about the return policy for this product. I purchased it a few weeks ago and it is defective.\"}\n",
+    "]"
```
Comment (lfayoux) [inline notebooks/sagemaker/Deploy rerank v3.5 model.ipynb]: documents should just be a list of strings

File: notebooks/sagemaker/Deploy rerank v3.5 model.ipynb [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "co = Client(region_name=region)\n",
+    "co.create_endpoint(arn=model_package_arn, endpoint_name=\"cohere-rerank-english-v3-0\", instance_type=\"ml.g5.xlarge\", n_instances=1)\n",  
+    "\n",
+    "# If the endpoint is already created, you just need to connect to it\n",
+    "# co.connect_to_endpoint(endpoint_name=\"cohere-rerank-english-v3-0\")"
```
Comment (lfayoux) [inline notebooks/sagemaker/Deploy rerank v3.5 model.ipynb]: 3.5?

File: notebooks/sagemaker/Deploy rerank v3.5 model.ipynb (hunk 1 above)
Comment (jpekmez) [reply notebooks/sagemaker/Deploy rerank v3.5 model.ipynb]: no it's draft, we still need to update the packages and rename to 3.5

File: notebooks/sagemaker/Deploy rerank v3.5 model.ipynb [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    "### A. Create an endpoint"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "co = Client(region_name=region)\n",
```
Comment (lfayoux) [inline notebooks/sagemaker/Deploy rerank v3.5 model.ipynb:140]: can we do ClientV2 instead here, we want to only show v2 examples for 3.5

File: notebooks/sagemaker/Deploy rerank v3.5 model.ipynb [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    "### C. Perform real-time inference"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "response = co.rerank(documents=documents, query='What emails have been about returning items?', rank_fields=[\"Title\",\"Content\"], top_n=5)"
```
Comment (lfayoux) [inline notebooks/sagemaker/Deploy rerank v3.5 model.ipynb]: no rank_fields pls

File: notebooks/sagemaker/Deploy rerank v3.5 model.ipynb [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    "1. Once you click on **Continue to configuration button** and then choose a **region**, you will see a **Product Arn** displayed. This is the model package ARN that you need to specify while creating a deployable model using Boto3. Copy the ARN corresponding to your region and specify the same in the following cell."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install --upgrade setuptools==69.5.1 cohere-aws\n",
```
Comment (lfayoux) [inline notebooks/sagemaker/Deploy rerank v3.5 model.ipynb:69]: pls use the normal cohere sdk instead

File: notebooks/sagemaker/Deploy rerank v3.5 model.ipynb (hunk 2 above)
Comment (mahjongmen) [reply notebooks/sagemaker/Deploy rerank v3.5 model.ipynb]: agreed

**Comment by mahjongmen** [comment]
@lfayoux @jpekmez 

### In the markdown of the notebook, change the following: 

- [cohere-rerank-v3.5](https://aws.amazon.com/marketplace/pp/prodview-rqhxjsjanb3gy) should be replaced with the new market listing
- or your AWS account has a subscription to [cohere-rerank-v3.5](https://aws.amazon.com/marketplace/pp/prodview-xwsyvhz7rkjqe) should be replaced with the new market listing
- The model package listing page should be updated to cohere-rerank-v3.5 listing page

### For the code: 
From the docs: https://docs.cohere.com/docs/cohere-works-everywhere#sagemaker it looks like we do not support v2 API yet @lfayoux which one do you want to use? I also looked at the client code itself here: https://github.com/cohere-ai/cohere-python/blob/main/src/cohere/sagemaker_client.py and it's not clear to me how the model arn packages work in this case

**Comment by lfayoux** [comment]
@mahjongmen There should be a new sdk release with sagemaker v2 client. Either was we should replace the sdk with the normal one

**Comment by mahjongmen** [comment]
@lfayoux when I spoke with @mkozakov yesterday it was determined that because @billytrend-cohere is gonna push the new SDK on monday morning but because @gloriacohere needs to push all of this monday morning at 6:30am EST it won't make it in time. 

Issue right now is that the cohere sdk is only compatible with V1 API

**Comment by mkozakov** [comment]
We can actually use the cohere sdk with client v1 and then later swap to v2

**Comment by mahjongmen** [comment]
great - can someone do that. I looked through our documents and it wasn't obvious to me whether or not we needed ARN etc... so I don't want to change the docs without testing

**Comment by mahjongmen** [comment]
as per https://cohereai.slack.com/archives/C07SBRVE8G2/p1733079402682219 we should go ahead with `cohere-aws` sdk and then change to using the cohere sdk during monday (post release)

File: notebooks/sagemaker/Deploy rerank v3.5 model.ipynb (hunk 5 above)
Comment (mahjongmen) [reply notebooks/sagemaker/Deploy rerank v3.5 model.ipynb:69]: to be resolved post monday

**Review by CoderHam — APPROVED** [review]

**Pull request closed without merging**